In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
%matplotlib inline

# helps in creating the graphs in the notebook otherwise they will pop up

In [3]:
data = pd.read_csv('student.csv')
display (data.head())

ID school sex  age address famsize Pstatus  Medu  Fedu     Mjob  ...  \
0  SD0001     GP   F   18       U     GT3       A     4     4  at_home  ...   
1  SD0002     GP   F   17       U     GT3       T     1     1  at_home  ...   
2  SD0003     GP   F   15       U     LE3       T     1     1  at_home  ...   
3  SD0004     GP   F   15       U     GT3       T     4     2   health  ...   
4  SD0005     GP   F   16       U     GT3       T     3     3    other  ...   

  internet romantic famrel  freetime  goout  Dalc Walc health absences Grade  
0       no       no    NaN         3      4     1    1      3      NaN     1  
1      yes       no    NaN         3      3     1    1      3      NaN     1  
2      yes       no    NaN         3      2     2    3      3     10.0     2  
3      yes      yes    3.0         2      2     1    1      5      2.0     3  
4       no       no    4.0         3      2     1    2      5      4.0     2  

[5 rows x 32 columns]

In [4]:
data.shape

(395, 32)

In [5]:
data.info()

# Null Count - missing value in the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 32 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          395 non-null    object 
 1   school      395 non-null    object 
 2   sex         395 non-null    object 
 3   age         395 non-null    int64  
 4   address     395 non-null    object 
 5   famsize     395 non-null    object 
 6   Pstatus     395 non-null    object 
 7   Medu        395 non-null    int64  
 8   Fedu        395 non-null    int64  
 9   Mjob        395 non-null    object 
 10  Fjob        395 non-null    object 
 11  reason      395 non-null    object 
 12  guardian    395 non-null    object 
 13  traveltime  395 non-null    int64  
 14  studytime   395 non-null    int64  
 15  failures    395 non-null    int64  
 16  schoolsup   395 non-null    object 
 17  famsup      395 non-null    object 
 18  paid        395 non-null    object 
 19  activities  311 non-null    o

In [6]:
data.columns

Index(['ID', 'school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu',
       'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'Grade'],
      dtype='object')

# Data Dictionary

Attribute Information ::

Attributes for both student-mat.csv (Math course) and student-por.csv (Portuguese language course) datasets: 
1.	school - student's school (binary: 'GP' - Gabriel Pereira or 'MS' - Mousinho da Silveira) 
2.	sex - student's sex (binary: 'F' - female or 'M' - male) 
3.	age - student's age (numeric: from 15 to 22) 
4.	address - student's home address type (binary: 'U' - urban or 'R' - rural) 
5.	famsize - family size (binary: 'LE3' - less or equal to 3 or 'GT3' - greater than 3) 
6.	Pstatus - parent's cohabitation status (binary: 'T' - living together or 'A' - apart) 
7.	Medu - mother's education (numeric: 0 - none, 1 - primary education (4th grade), 2 - 5th to 9th grade, 3 - secondary education or 4 - higher education) 
8.	Fedu - father's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education) 
9.	Mjob - mother's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other') 
10.	Fjob - father's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other') 
11.	reason - reason to choose this school (nominal: close to 'home', school 'reputation', 'course' preference or 'other') 
12.	guardian - student's guardian (nominal: 'mother', 'father' or 'other') 
13.	traveltime - home to school travel time (numeric: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour) 
14.	studytime - weekly study time (numeric: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours) 
15.	failures - number of past class failures (numeric: n if 1<=n<3, else 4) 
16.	schoolsup - extra educational support (binary: yes or no) 
17.	famsup - family educational support (binary: yes or no) 
18.	paid - extra paid classes within the course subject (Math or Portuguese) (binary: yes or no) 
19.	activities - extra-curricular activities (binary: yes or no) 
20.	nursery - attended nursery school (binary: yes or no) 
21.	higher - wants to take higher education (binary: yes or no) 
22.	internet - Internet access at home (binary: yes or no) 
23.	romantic - with a romantic relationship (binary: yes or no) 
24.	famrel - quality of family relationships (numeric: from 1 - very bad to 5 - excellent) 
25.	freetime - free time after school (numeric: from 1 - very low to 5 - very high) 
26.	goout - going out with friends (numeric: from 1 - very low to 5 - very high) 
27.	Dalc - workday alcohol consumption (numeric: from 1 - very low to 5 - very high) 
28.	Walc - weekend alcohol consumption (numeric: from 1 - very low to 5 - very high) 
29.	health - current health status (numeric: from 1 - very bad to 5 - very good) 
30.	absences - number of school absences (numeric: from 0 to 93) 

## isnull()
- The isnull() function is used to detect missing values for an array-like object.

- This function takes a scalar or array-like object and indicates whether values are missing (NaN in numeric arrays)

In [7]:
data.isnull()

# it will only check for NAN and not BLANK VALUES
# TRUE = NAN = Missing Value

ID  school    sex    age  address  famsize  Pstatus   Medu   Fedu  \
0    False   False  False  False    False    False    False  False  False   
1    False   False  False  False    False    False    False  False  False   
2    False   False  False  False    False    False    False  False  False   
3    False   False  False  False    False    False    False  False  False   
4    False   False  False  False    False    False    False  False  False   
..     ...     ...    ...    ...      ...      ...      ...    ...    ...   
390  False   False  False  False    False    False    False  False  False   
391  False   False  False  False    False    False    False  False  False   
392  False   False  False  False    False    False    False  False  False   
393  False   False  False  False    False    False    False  False  False   
394  False   False  False  False    False    False    False  False  False   

      Mjob  ...  internet  romantic  famrel  freetime  goout   Dalc   Walc  \
0    False  ...     False     False    True     False  False  False  False   
1    False  ...     False     False    True     False  False  False  False   
2    False  ...     False     False    True     False  False  False  False   
3    False  ...     False     False   False     False  False  False  False   
4    False  ...     False     False   False     False  False  False  False   
..     ...  ...       ...       ...     ...       ...    ...    ...    ...   
390  False  ...     False     False   False     False  False  False  False   
391  False  ...     False     False    True     False  False  False  False   
392  False  ...     False     False   False     False  False  False  False   
393  False  ...     False     False    True     False  False  False  False   
394  False  ...     False     False   False     False  False  False  False   

     health  absences  Grade  
0     False      True  False  
1     False      True  False  
2     False     False  False  
3     False     False  False  
4     False     False  False  
..      ...       ...    ...  
390   False     False  False  
391   False     False  False  
392   False     False  False  
393   False     False  False  
394   False     False  False  

[395 rows x 32 columns]

In [8]:
data.isnull().sum() # col wise sum ===  T = 1 and F = 0

ID              0
school          0
sex             0
age             0
address         0
famsize         0
Pstatus         0
Medu            0
Fedu            0
Mjob            0
Fjob            0
reason          0
guardian        0
traveltime      0
studytime       0
failures        0
schoolsup       0
famsup          0
paid            0
activities     84
nursery         0
higher          0
internet        0
romantic        0
famrel        117
freetime        0
goout           0
Dalc            0
Walc            0
health          0
absences       58
Grade           0
dtype: int64

## Deletion

In [10]:
data_del = data.dropna(how = "any") # any row that has a missing value drop it

In [11]:
print ('Before: ', data.shape, 'After: ', data_del.shape)

Before:  (395, 32) After:  (182, 32)


In [12]:
395 - 182

213

In [13]:
round(213/395, 2)

0.54

**We will end up deleting 54% of data.**

**We will lose 54% of the data, this will hamper the analysis, since we require large amount of data for Predictive Analysis.** 